File: leetcode_ly/c++/Array/MajorityElementTwo.cpp
class Solution {
public:
    vector<int> majorityElement(vector<int>& nums) {
        vector<int> val;
        vector<int> cnt;
        vector<int> res;
        if(nums.size() == 0) return res;
        for(int i = 0; i < nums.size(); i++){
            if(val.size() == 0){
                val.push_back(nums[i]);
                cnt.push_back(1);
            }
            else if(val.size() == 1){
                if(val[0] == nums[i]) cnt[0]++;
                else{
                    val.push_back(nums[i]);
                    cnt.push_back(1);
                }
            }
            else{
                if(val[0] == nums[i]) cnt[0]++;
                else if(val[1] == nums[i]) cnt[1]++;
                else{
                    cnt[0]--;
                    cnt[1]--;
                    if(cnt[1] == 0){
                        val.pop_back();
                        cnt.pop_back();
                    } 
                    if(cnt[0] == 0){
                        val.erase(val.begin());
                        cnt.erase(cnt.begin());
                    }
                }
            }
        }
        
        for(int i = 0; i < val.size(); i++){
            int temp = 0;
            for(int j = 0; j < nums.size(); j++){
                if(val[i] == nums[j]) temp++;
            }
            if(temp > nums.size() / 3) res.push_back(val[i]);
        }
        return res;
    }
};

File: leetcode_ly/java/Array/MajorityElementTwo.java
public class Solution {
    public List<Integer> majorityElement(int[] nums) {
        List<Integer> res = new ArrayList<Integer>();
        if(nums.length == 0) return res;
        int num1 = nums[0];
        int num2 = nums[0];
        int cnt1 = 0;
        int cnt2 = 0;
        
        for(int i = 0; i < nums.length; i++){
            if(num1 == nums[i]) cnt1++;
            else if(num2 == nums[i]) cnt2++;
            else if(cnt1 == 0){
                num1 = nums[i];
                cnt1 = 1;
            }
            else if(cnt2 == 0){
                num2 = nums[i];
                cnt2 = 1;
            }
            else{
                cnt1--;
                cnt2--;
            }
        }
        
        cnt1 = 0;
        cnt2 = 0;
        for(int i = 0; i < nums.length; i++){
            if(nums[i] == num1) cnt1++;
            //if(nums[i] == num2 && num1 != num2) cnt2++;
            else if(nums[i] == num2) cnt2++;
        }
        if(cnt1 > nums.length / 3) res.add(num1);
        if(cnt2 > nums.length / 3) res.add(num2);
        return res;
    }
}

map